### PR TITLE
Fixed multichannel bug.

### PR DESCRIPTION
**Pull request by katmsft**

### CODE REVIEW

File: sdk/storage/azure-storage-files-shares/test/share_service_client_test.cpp [hunk 1]
```diff
@@ -146,6 +146,8 @@ namespace Azure { namespace Storage { namespace Test {
   TEST_F(FileShareServiceClientTest, SetProperties)
   {
     auto properties = *m_fileShareServiceClient->GetProperties();
+    // Has to remove before set, otherwise would return failure.
+    properties.Protocol = Core::Nullable<Files::Shares::ShareProtocolSettings>();
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-files-shares/test/share_service_client_test.cpp:150]: I don't understand this comment. Can you explain the bug please.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp [hunk 2]
```diff
@@ -1898,9 +1889,9 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
             {
               path.emplace_back(XmlTagName::c_StorageServiceProperties);
             }
-            else if (std::strcmp(node.Name, "Protocol") == 0)
+            else if (std::strcmp(node.Name, "ProtocolSettings") == 0)
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1892]: What motivated this rename?

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp [hunk 3]
```diff
@@ -1406,7 +1409,9 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
 
       static void ShareProtocolSettingsToXml(XmlWriter& writer, const ShareProtocolSettings& object)
       {
+        writer.Write(XmlNode{XmlNodeType::StartTag, "ProtocolSettings"});
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1412]: Add unit tests to capture regressions for such issues.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp [hunk 4]
```diff
@@ -1772,11 +1774,7 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
           else if (node.Type == XmlNodeType::StartTag)
           {
 
-            if (std::strcmp(node.Name, "SMB") == 0)
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1775]: Why don't we need to check for this anymore? Also, add tests.

File: sdk/storage/azure-storage-files-shares/test/share_service_client_test.cpp (hunk 1 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/test/share_service_client_test.cpp:150]: Currently, the service is rejecting `ProtocolSettings` to be set for subscriptions that haven't turned this feature on, although this field is returned in the response body.
We are trying to communicate with the service team if this is correct behavior, but in the meantime, this line needs to exist.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp (hunk 2 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1892]: A failed unit test. The unit test is disabled due to service not ready but failed in a test environment that has this feature enabled.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp (hunk 3 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1412]: This is already covered in `DISABLED_SetPremiumFileProperties`. However, the service is not ready to enable the test in the production environment.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp (hunk 4 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1775]: Previously there was a bug adding a redundant `SMB` tag parsing, which is already done in `ShareProtocolSettingsFromXml`.

Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1775]: This is again covered in `DISABLED_SetPremiumFileProperties`. This test is manually run against a test environment but disabled due to production not ready.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp [hunk 5]
```diff
@@ -1338,7 +1338,8 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
       static void ShareRetentionPolicyToXml(XmlWriter& writer, const ShareRetentionPolicy& object)
       {
         writer.Write(XmlNode{XmlNodeType::StartTag, "Enabled"});
-        writer.Write(XmlNode{XmlNodeType::Text, nullptr, object.Enabled == 0 ? "false" : "true"});
+        writer.Write(
+            XmlNode{XmlNodeType::Text, nullptr, object.Enabled == false ? "false" : "true"});
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1342]:
```suggestion
            XmlNode{XmlNodeType::Text, nullptr, object.Enabled? "true" : "false"});
```

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp [hunk 6]
```diff
@@ -1431,9 +1436,7 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
         }
         if (object.Protocol.HasValue())
         {
-          writer.Write(XmlNode{XmlNodeType::StartTag, "Protocol"});
           ShareProtocolSettingsToXml(writer, object.Protocol.GetValue());
-          writer.Write(XmlNode{XmlNodeType::EndTag});
         }
         writer.Write(XmlNode{XmlNodeType::EndTag});
       };
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1442]: do we need `;` here?

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp [hunk 7]
```diff
@@ -1511,7 +1514,7 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
           {
             if (path.size() == 1 && path[0] == XmlTagName::c_Enabled)
             {
-              result.Enabled = node.Value;
+              result.Enabled = (node.Value == std::string("true"));
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1517]: You're using `std::strcmp`  in the code below, be consistent.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp [hunk 8]
```diff
@@ -1898,9 +1889,9 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
             {
               path.emplace_back(XmlTagName::c_StorageServiceProperties);
             }
-            else if (std::strcmp(node.Name, "Protocol") == 0)
+            else if (std::strcmp(node.Name, "ProtocolSettings") == 0)
             {
-              path.emplace_back(XmlTagName::c_Protocol);
+              path.emplace_back(XmlTagName::c_ProtocolSettings);
             }
             else if (std::strcmp(node.Name, "Cors") == 0)
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1896]:
```suggestion
            else if (std::strcmp(node.Name, XmlTagName::c_Cors) == 0)
```

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp (hunk 2 above)
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1892]:
```suggestion
            else if (std::strcmp(node.Name, XmlTagName::c_ProtocolSettings) == 0)
```

Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1892]: be consistent

File: sdk/storage/azure-storage-files-shares/test/share_service_client_test.cpp (hunk 1 above)
Comment (Jinming-Hu) [reply sdk/storage/azure-storage-files-shares/test/share_service_client_test.cpp:150]: Then create an issue for this.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp (hunk 8 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1896]: `XmlTagName::c_Cors` is the enum used to identify the current level of XML.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp (hunk 2 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1892]: `XmlTagName::c_Cors` is the enum used to identify the current level of XML.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp (hunk 6 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1442]: We do not, removed.

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp (hunk 7 above)
Comment (katmsft) [reply sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/protocol/share_rest_client.hpp:1517]: Nice catch. Resolved in the next commit.

**Review by Jinming-Hu — APPROVED** [review]

**Comment by ahsonkhan** [comment]
Hey @Jinming-Hu, for your **initial** review of a PR, please consider "Starting a review" to group all your feedback rather than leaving individual comment. Otherwise, there is a flood of email/notifications for each comment rather than single notification with all feedback. Thanks!
image

image

For post replies/threads or a few/one-off comments, standalone comments are fine.

**Pull request merged**